kernel/src/dynamic_process_loading.rs

```diff
@@ -0,0 +1,1242 @@
+                if proc_index < new_process_count {
+                    let process_start_address = value.unwrap().get_addresses().flash_start;
+                    let process_end_address = value.unwrap().get_addresses().flash_end;
+
+                    if new_process_end_address >= process_start_address
+                        && new_process_end_address <= process_end_address
+                    {
+                        /* Case 1
+                         *              _________________          _______________           _________________
+                         *  ___________|__               |        |              _|_________|__               |
+                         * |           |  |              |        |             | |         |  |              |
+                         * |   new app |  |  app2        |   or   |   app1      | | new app |  |  app2        |
+                         * |___________|__|              |        |             |_|_________|__|              |
+                         *             |_________________|        |_______________|         |_________________|
+                         *
+                         * ^...........^                                           ^........^
+                         * In this case, we discard this region and try to find another start address from the end address + 1 of app2
+                         */
+                        return Err((process_end_address + 1, ProcessLoadError::NotEnoughMemory));
+                    } else if new_process_start_address >= process_start_address
+                        && new_process_start_address <= process_end_address
+                    {
+                        /* Case 2
+                         *              _________________
+                         *  ___________|__               |    _______________
+                         * |           |  |              |   |               |
+                         * |   app2    |  |  new app     |   |     app3      |
+                         * |___________|__|              |   |_______________|
+                         *             |_________________|
+                         *
+                         *                 ^
+                         *                 | In this case, the start address of new app is replaced by 'the end address + 1' of app2,
+                         *                   and we try to find another start address from the end address + 1 of app2 and recheck for
+                         *                   the previous condition
+                         */
+                        return Err((process_end_address + 1, ProcessLoadError::NotEnoughMemory));
+                    }
+                }
+            }
+            Ok(())
+        });
+        Ok(())
+    }
+
+    fn find_next_available_address(
+        &self,
+        flash: &'static [u8],
+        app_length: usize,
+    ) -> Result<usize, ErrorCode> {
```

Comment:
I think it would more straightforward to just have this function return more, and encompass all of the work of determining where an app can go and what padding requirements there are. So this would return the address and the padding needs.

Reply:
I think padding needs should not be returnable because we are setting it using an OptionalCell. This way, we don't have to call the function again during load to check if we need to write pre-pad. (I know I am calling it right now, I should not be doing that because the value is set the first time anyway, so I removed the check_padding_req() from load).